doc/designs/application-specific-passwords.md

```diff
@@ -0,0 +1,283 @@
+* two-component principals with a host as the first component are always host
+  principals (`_host_/_fqdn_`)
+* all other two-component principals are always service principals
+  (`_service_/_fqdn_`)
+* single-component principals are always users (user)
+Now, they will be:
+* two-component principals with an application name as the first component are
+  users logged in with an application-specific password (`_appname_/_user_`)
+* two-component principals with a host as the first component are host
+  principals (`_host_/_fqdn_`)
+* all other two-component principals are service principals (`_service_/_fqdn_`)
+* single-component principals are users logged in with their primary password
+  (`user`)
+
+### Permissions management for application-specific passwords
+
+New ACIs must automatically be added and maintained for every
+application-specific password so that the owning user will have read-write
+access to all (and only) objects holding his application-specific passwords - 
+both when logged in using the primary password and when logged in using the
+respective application-specific password. This is the only thing to do with
+respect to RBAC and application-specific passwords, as RBAC serves only for work
+with LDAP in FreeIPA.
+
+The only general limitation for subaccounts will be the inability to manage the
+account (modify the information about the user, change his other passwords...).
+As all permissions for them will be assigned by administrators separately and
+from scratch, even a limitation such as reducing the effective permissions of
+the current user session (possibly according to the application policy) using a
+389-DS plugin probably will not be needed.
+
+#### HBAC for application-specific passwords
+
+Two-component principals for subaccounts, as described above, will be used for
+assigning HBAC rules to users logged in using application-specific passwords.
+These HBAC rules will be assigned by administrators just as the other ones.
+
+TBD: When an application-specific password is deleted, delete the associated
+HBAC rules automatically, or keep them?
+
+Optionally, a possibility for administrators to specify that users can log into
+a particular service only using the passwords from a specified LDAP subtree (or
+the primary password) can be provided.
+
+hbacrule plugin therefore will need to be extended with the ability to assign
+allowed LDAP subtrees to application-specific passwords as a part of HBAC rules.
+hbactest plugin will be extended to test this, and it will be used to test HBAC
+with application-specific passwords in general.
+
+### SSSD support for the aforementioned functionality
+
+SSSD will also be extended to support the aforementioned functionality. This
+will be done by a closely related, but separate extension of SSSD. On the
+side of FreeIPA, after SSSD is extended, interactions with SSSD using the new
+functionality must be tested in `ipatests/test_integration/test_sssd.py`.
+
```

Comment:
What kind of modifications are envisioned for SSSD ?
I can't think of any reason to modify SSSD atm, but perhaps there are some assumptions here not spelled out that I am not thinking of.
Please elaborate.

Reply:
Is this to allow app-specific password to be used via PAM?
Pondering if this should actually be allowed...

Reply:
I don't know yet, @stlaz told me that it will probably needed, and that I should consult it with SSSD upstream. Now I concentrate on extending FreeIPA, and according to that we will decide what - if anything - needs to be done with SSSD.